app/src/ble.c

```diff
@@ -612,32 +622,69 @@ int zmk_ble_unpair_all() {
     return resp;
 };
 
-bool zmk_ble_handle_key_user(struct zmk_key_event *key_event) {
-    zmk_key_t key = key_event->key;
+#if IS_ENABLED(CONFIG_ZMK_BLE_PASSKEY_ENTRY)
+static int zmk_ble_handle_key_user(struct zmk_keycode_state_changed *event) {
+    zmk_key_t key = event->keycode;
+
+    LOG_DBG("key %d", key);
 
     if (!auth_passkey_entry_conn) {
-        return true;
+        LOG_DBG("No connection for passkey entry");
+        return ZMK_EV_EVENT_BUBBLE;
+    }
+
+    if (!event->state) {
+        LOG_DBG("Key released, ignoring");
+        return ZMK_EV_EVENT_BUBBLE;
+    }
+
+    if (key == HID_USAGE_KEY_KEYBOARD_ESCAPE) {
+        bt_conn_auth_cancel(auth_passkey_entry_conn);
+        return ZMK_EV_EVENT_HANDLED;
     }
 
-    if (key < NUMBER_1 || key > NUMBER_0) {
-        return true;
+    if (key < HID_USAGE_KEY_KEYBOARD_1_AND_EXCLAMATION ||
```

Comment:
Is it worth checking for numpad digits as well? Not a big deal either way, but I can see folks trying to use them on a 100% keyboard.

Reply:
Perhaps. The host OS doesn't treat them as numeric unless caps lock is enabled, but I suppose we could ignore that requirement for our use.

Reply:
@caksoylar Added.